api/src/main/java/org/openmrs/module/htmlformentryui/element/ObsFromFragmentElement.java

```diff
@@ -0,0 +1,200 @@
+				sb.append(fragmentParams.get("defaultDate"));
+			}
+			if (FragmentType.DROPDOWN == fragmentType) {
+				sb.append(fragmentParams.get("defaultDate"));
+			}
+			sb.append("</span></p>");
+			return sb.toString();
+		}
+		try {
+			return uiUtils.includeFragment(viewProvider, fragment, fragmentParams);
+		} catch (NullPointerException e) {
+			// if we are validating/submitting the form, then this method is being called from a fragment action method
+            // and the UiUtils we have access to doesn't have a FragmentIncluder. That's okay, because we don't actually
+            // need to generate the HTML, so we can pass on this exception.
+            // (This is hacky, but I don't see a better way to do it.)
+            return "Submitting the form, so we don't generate HTML";
+		} catch (PageAction e) {
+			throw new IllegalStateException("Tried to include a fragment that threw a PageAction.", e);
+		}
+	}
+	
+	private Object convertToType(String val) {
+		ConceptDatatype dataType = concept.getDatatype();
+		if (dataType.isDate()) {
+			return HtmlFormEntryUtil.convertToType(val, Date.class);
+		}
+		if (dataType.isDateTime() || dataType.isTime()) {
+			try {
+				DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+				df.setLenient(false);
+				return df.parse(val);
+			} catch (ParseException e) {
+				throw new IllegalArgumentException(e);
+			}
+		}
+		if (dataType.isCoded()) {
+			for (ConceptAnswer answer : concept.getAnswers()) {
+				if (answer.getAnswerConcept().getName().getName().equals(val)) {
+					return answer.getAnswerConcept();
+				}
+			}
+		}
+		return val;
+	}
+
+	private void setFragmentTypeFromFragmentName(String fragment) {
+		if (fragment.contains("/")) {
+			fragment = fragment.split("/")[1];
+		}
+		try {
+			fragmentType = FragmentType.valueOf(fragment.toUpperCase());
+	
+		} catch(IllegalArgumentException e) {
+			// Ignore
+		}
+	}
+	
+	private List<String> getConceptAnswersNames(Concept question) {
+		List<String> ret = new ArrayList<String>();
+		for (ConceptAnswer candidate : question.getAnswers()) {
+			ret.add(candidate.getConcept().getName().getName());
+		}
+		return ret;
+	}
+	
+	private void setupFragmentParams(FormEntryContext context) {
+		if (FragmentType.DROPDOWN == fragmentType) {
+			fragmentParams.put("options", getConceptAnswersNames(concept));
+		}
+		if (FormEntryContext.Mode.EDIT == context.getMode() || FormEntryContext.Mode.VIEW == context.getMode()) {
+			Encounter encounter = context.getExistingEncounter();
+			if (encounter != null) {
+				for (Obs candidate : encounter.getObsAtTopLevel(false)) {
+					if (candidate.getConcept().equals(concept)) {
+						if (FragmentType.DATETIMEPICKER == fragmentType) {
+							fragmentParams.put("defaultDate", candidate.getValueDatetime());
+						}
+						if (FragmentType.DROPDOWN == fragmentType) {
+							fragmentParams.put("initialValue", candidate.getValueCoded().getName().getName());
+						}
+						break;
```

Comment:
The whole logic above will have to be changed. There's two things:

1. What value should be extracted from `candidate`?
This is not driven by the type of fragment but **by the concept datatype**.
2. Which fragment param should be set to the value extracted from `candidate`?
This should be parameterizable. Maybe we need another attribute for this, like `initFragmentParamName`.

Eg.:
```html
<obsFromFragment
 conceptId="..."
 provider="..."
 fragment="..."
 fragmentParams="formFieldName=foo..."
 initFragmentParamName="initialValue"
 ... />
```